### ISSUE TITLE
is-obj and is-object merged incorrectly

### ISSUE THREAD

**Issue by daveisfera** [yarnpkg/yarn#4737]
**Do you want to request a *feature* or report a *bug*?**
Bug report

**What is the current behavior?**
`is-obj` and `is-object` were merged incorrectly but they should be kept separate.

**If the current behavior is a bug, please provide the steps to reproduce.**
`is-obj` was changed like this in one branch:
```
-is-obj@^1.0.1:
+is-obj@^1.0.0, is-obj@^1.0.1:
   version "1.0.1"
   resolved "https://registry.yarnpkg.com/is-obj/-/is-obj-1.0.1.tgz#3e4729ac1f5fde025cd7d83a896dab9f4f67db0f"
```

Then `is-object` was added in another branch:
```
+is-object@~1.0.1:
+  version "1.0.1"
+  resolved "https://registry.yarnpkg.com/is-object/-/is-object-1.0.1.tgz#8952688c5ec2ffd6b03ecc85e769e02903083470"
+
```

Then the merge from git resulted in the following conflict:
```
<<<<<<< HEAD
is-obj@^1.0.0, is-obj@^1.0.1:
||||||| merged common ancestors
is-obj@^1.0.1:
=======
is-object@~1.0.1:
>>>>>>> qa
  version "1.0.1"
  resolved "https://registry.yarnpkg.com/is-object/-/is-object-1.0.1.tgz#8952688c5ec2ffd6b03ecc85e769e02903083470"
```

I would have expected the `resolved` line to also be included in the conflict, but for some reason that wasn't the case.

**What is the expected behavior?**
Both `is-obj` and `is-object` are maintained in the `yarn.lock` file.

**Please mention your node.js, yarn and operating system version.**
node 8.7.0
yarn 1.2.0
Alpine Docker image

**Comment by also** [comment]
I had the same issue () with `strip-bom-stream` and `strip-bom`. It seems like there's a prefix comparison that's matching things it shouldn't

**Comment by gnapse** [comment]
I'm having the same issue with `babel-core`. I'm adding it to a project because of a peer dependency of another stuff I added, and I get this error message:

```
$ yarn add -D babel-core
yarn add v1.3.2
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
warning Pattern ["babel-core@^6.26.0"] is trying to unpack in the same destination "/Users/ernesto/Library/Caches/Yarn/v1/npm-babel-core-6.26.0-af32f78b31a6fcef119c87b0fd8d9753f03a0bb8" as pattern ["babel-core@^6.0.0","babel-core@^6.26.0","babel-core@6.26.0"]. This could result in a non deterministic behavior, skipping.
```

**Comment by Dylanlan** [comment]
I seem to have ran into this issue today.
I had merge conflicts in my yarn.lock file, and I tried using `yarn install` to fix them.
It says it succeeded, but it had merged the `chance` and `character-reference-invalid` package entries together, without a resolved value for chance, like this:

```
chance@1.1.3, chance@^1.0.9, character-reference-invalid@^1.0.0:
  version "1.1.3"
  resolved "https://registry.yarnpkg.com/character-reference-invalid/-/character-reference-invalid-1.1.3.tgz#1647f4f726638d3ea4a750cf5d1975c1c7919a85"
  integrity sha512-VOq6PRzQBam/8Jm6XBGk2fNEnHXAdGd6go0rtd4weAGECBamHDwwCQSOT12TACIYUZegUXnV6xBXqUssijtxIg==
```

I ended up fixing it by manually copying both package entries from git history, which wasn't ideal.

My versions:
macOS Catalina 10.15.1
yarn 1.19.1
node 10.17.0

It's interesting that it seems to be related to packages with common prefixes, and in my case a prefix of only 3 characters.
I might try to look into this further if I have time.
